### PR TITLE
feat(dart): add graph support

### LINKED ISSUE #5366: feat(dart): add dependency tree support
## Description
Dart uses `package_config.json` file (under the `.dart_tool/ directory`) - https://dart.dev/guides/packages#getting-packages
We can get path to dependency `pubspec.yaml` file and get list of child dependencies.

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
[Cache dir](https://dart.dev/tools/pub/glossary#system-cache) contains `pubspec.yaml` for each dependency.
We can parse these files and use them to build dependency graph.

example of work:
```
➜ PUB_CACHE=/Users/work/work/temp/5366/app/cache ./trivy -d fs --dependency-tree ./pubspec.lock   
...
2023-10-16T13:24:43.488+0600	INFO	Number of language-specific files: 1
2023-10-16T13:24:43.488+0600	INFO	Detecting pub vulnerabilities...
2023-10-16T13:24:43.488+0600	DEBUG	Detecting library vulnerabilities, type: pub, path: pubspec.lock

pubspec.lock (pub)

Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 1, CRITICAL: 0)

┌─────────┬────────────────┬──────────┬────────┬───────────────────┬───────────────┬──────────────────────────────────────────────────────────┐
│ Library │ Vulnerability  │ Severity │ Status │ Installed Version │ Fixed Version │                          Title                           │
├─────────┼────────────────┼──────────┼────────┼───────────────────┼───────────────┼──────────────────────────────────────────────────────────┤
│ dio     │ CVE-2021-31402 │ HIGH     │ fixed  │ 4.0.6             │ 5.0.0         │ dio vulnerable to CRLF injection with HTTP method string │
│         │                │          │        │                   │               │ https://avd.aquasec.com/nvd/cve-2021-31402               │
└─────────┴────────────────┴──────────┴────────┴───────────────────┴───────────────┴──────────────────────────────────────────────────────────┘

Dependency Origin Tree (Reversed)
=================================
pubspec.lock
└── dio@4.0.6, (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 1, CRITICAL: 0)
    └── retrofit@2.0.0
```

TODO for another PR:
- add support of `Package.Dev` flag (to exclude dev deps)
- add parsing of `LICENSE` files (we can check `LICENSE` file next to `pubspec.yaml` and find licenses)
- ??? use `:` separator for ID - https://dart.dev/tools/pub/cmd/pub-add#version-constraint

## Related issues
- Close #5366 

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/fanal/analyzer/language/dart/pub/pubspec.go [hunk 1]
```diff
@@ -2,35 +2,177 @@ package pub
...
+	if err = yaml.NewDecoder(f).Decode(&spec); err != nil {
+		return "", nil, xerrors.Errorf("unable to decode %q: %w", filePath, err)
+	}
+	if len(spec.Dependencies) > 0 {
+		// pubspec.yaml uses version ranges
+		// save only dependencies names
+		dependsOn := lo.MapToSlice(spec.Dependencies, func(key string, _ interface{}) string {
+			return key
+		})
+		return utils.PackageID(spec.Name, spec.Version), dependsOn, nil
```
Comment (knqyf263) [inline pkg/fanal/analyzer/language/dart/pub/pubspec.go:173]: It is better to align the happy path to the left.

```suggestion
	if len(spec.Dependencies) == 0 {
		return "", nil, nil
```

File: pkg/fanal/analyzer/language/dart/pub/pubspec.go [hunk 2]
```diff
@@ -2,35 +2,177 @@ package pub
...
-	res, err := language.Analyze(types.Pub, input.FilePath, input.Content, p)
+func findDependsOn() (map[string][]string, error) {
+	dir := cacheDir()
+	if !fsutils.DirExists(dir) {
+		log.Logger.Debugf("Cache dir (%s) not found. Need 'dart pub get' to fill dependency relationships", dir)
+		return nil, nil
+	}
+
+	deps := make(map[string][]string)
+	if err := filepath.WalkDir(dir, func(p string, d fs.DirEntry, err error) error {
```
Comment (knqyf263) [inline pkg/fanal/analyzer/language/dart/pub/pubspec.go:108]: Can't we use fsutils.WalkDir?

Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/dart/pub/pubspec.go:108]: This is same case as with `go.mod` licenses - https://github.com/aquasecurity/trivy/blob/6040d9f43a35819d4c0338839b97619338e753ed/pkg/fanal/analyzer/language/golang/mod/mod.go#L265

Some example:
User uses default cache path (`~/.pub_cache`).
But he scans project dir (e.g. `~/project1`).
In this case we don't add `pubspec.yaml` files from `~/.pub_cache` in FS for PostAnalyze. 
=> we can't read these files from `input.FS`.
Correct me, if i missed something.

Perhaps we can use `os.DirFS("/")` (`fsutils.WalkDir(os.DirFS("/"), ".", required, func(path string, d fs.DirEntry, r io.Reader) error `). But i am not sure that it make sense.

File: pkg/fanal/analyzer/language/dart/pub/pubspec.go (hunk 1 above)
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/dart/pub/pubspec.go]: Fixed in https://github.com/aquasecurity/trivy/pull/5374/commits/ffe7308f7258d614b98ec9752b01889944feb628

File: docs/docs/configuration/reporting.md [hunk 3]
```diff
@@ -65,6 +65,8 @@ The following packages/languages are currently supported:
     - Composer
 - Java
     - Maven: pom.xml
+- Dart
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply docs/docs/configuration/reporting.md]: Added tables in https://github.com/aquasecurity/trivy/pull/5374/commits/f5bbf4b55ed621341c63b3b293b343739e17e08d

File: pkg/fanal/analyzer/language/dart/pub/pubspec.go (hunk 2 above)
Comment (knqyf263) [reply pkg/fanal/analyzer/language/dart/pub/pubspec.go:108]: Doesn't `os.DirFS(dir)` work?

Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/dart/pub/pubspec.go]: After my suggestion about `os.DirFS` I didn't think of an obvious solution 😄 .
Thank you!
I used `fsutils.WalkDir` in https://github.com/aquasecurity/trivy/pull/5374/commits/3abdc473a0e6dc558969ea67c85348c03cbd7910

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**